[agl][guppy] Remove Qt usage and support building with CMake

nickdiego opened the pull request:
This PR is part of **Qt-less WAM** effort. Here we get rid of Qt usage in AGL's port of WAM on top of latest `@6.agl.guppy` branch, which is based on webosose's `chromium68` port. Additionally, it CMake port of WAM build system.

### Refactoring status:
- [x] ~~QMap/QList/QPair -> STL containers~~
- [x] ~~QString -> std::string~~
- [x] ~~QStringList -> std::vector\<std::string\>~~
- [x] ~~QFile/QFileInfo/QDir -> boost::filesystem~~
- [x] ~~QUrl -> custom Url helper class (initially GLib-based)~~
- [x] ~~QVariant removal: Some unused removed + other modified to use std::string~~
- [x] ~~QLocale removal: implemented function similar to QLocale::bcp47Name()~~
- [x] ~~Port build system to CMake~~

### Verification status:
- [x] ~~AGL Build~~ (@nickdiego)
- [x] ~~AGL validation on RPi: No regressions detected so far~~ (@nickdiego)
- [x] ~~AGL validation on RCarM3 and Minnowboard (@jaragunde / @ltilve)~~
- [ ] WebOS Build (@jdapena?)
- [ ] WebOS regression tests (@jdapena / @jaragunde)

#### Some remarks:
- This PR focus on common and AGL-specific code. WebOS-specific bits have been kept as much untouched as possible. So, that part might be addressed in a separate PR.
- To get this branch building correctly, a modified version of **[meta-agl-devel](https://github.com/nickdiego/meta-agl-devel/tree/qtless-cmake)** which adds _boost_ and removes _qtbase_ from WAM dependencies. A PR will be sent to AGL's meta-agl-devel repo once this one is merged.

[[SPEC-1911] Qt-less WAM: replace qt basic data structures](https://jira.automotivelinux.org/browse/SPEC-1911)
[[SPEC-1914] Qt-less WAM: move from QMake to CMake](https://jira.automotivelinux.org/browse/SPEC-1914)

nickdiego committed:
[agl][qtless] Replace Qt data container types by their STL counterpars

- This patch replaces Qt data structures, such as QMap, QPair, QList
by std::map, std::pair, and std::list/vector/set where applicable.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Migrate ApplicationDescription away from QString

This patch get rid of QString/QStringList usage from
ApplicationDescription implementation and some related APIs.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Add StringUtils initial implementation

StringUtils is intended to fill the gaps between
C++ standard library and Qt API's regarding string
manipulation. E.g: Qt provides a split() function
out-of-the-box while std library does not.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Get rid of QString usage in WebProcessManager

- This patch removes QString usage from WebProcessManager implementation.
To do that it also modifies classes that have some relationship with it,
such as WebAppManager and WebAppManagerService.
- Also, add some FIXME comments to mark points for next steps on removing
QString references from WAM.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Migrate WebAppManager away from QString

- This patch removes QString usage from WebAppManager, focusing
on removing its usage almost entirely from its public API, to do
that it also modifies classes that have some relationship with it,
such as WebApp(Base and Wayland), WebAppManagerService, WebAppFactory,
etc.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Do not use QString WebAppManager in JS strings building

This patch focus on removing QString usage in Javascript code generation
in WebApp initialization, also touching WebPageBase and WebPageBlink
implementation, modifying its 'evaluateJavascript()' method to receive
an std::string instead of QString.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] QString-less WebAppManager: SystemLanguage and DeviceInfo

This patch removes QString usage from SystemLanguage and DeviceInfo
API's in WebAppManager, WebAppManagerService, DeviceInfo, WebPageBase
and WebPageBlink.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] QString-less WebApp/WebPage: appId

This patch removes QString usage from appId APIs in WebApp and WebPage
classes. Devided to keep this as a separate patch because appId() heavy
usage all over WAM code.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] QString-less WebApp: Remove leftovers

Remove QString leftovers from WebAppBase and WebAppWayland,
besides that modifies interfaces from related classes.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Migrate WebAppWayland away from QVariant

This patch refactors WebAppWayland::setWindowProperty() method, removing
its dependency on QVariant. As most of its used was using QString and in
the end the value gets convered to a std::string the solution adopted
was to modify its parameter to a std::string.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Do not use QString in WebPage

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Do not use QString in NetworkStatus

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Do not use QString in WebAppManagerConfig

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Migrate PalmSystem away from QString

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] QString-less WAM: Remove leftovers

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Add Url parsing code

- The main goal of this patch is to remove WAM dependency on QUrl
implementation for Url Parsing and convenience API. To do that, we're
adding our own Url API, which in this initial version leverages GLib URI
Functions [1] for basic URI parsing and escaping.

- Also, as the code handling URLs in WAM is mostly tied to
QFile/QDir/QFileInfo manipulation, this patch also refactor that to use
Boost Filesystem module instead. It's worth noting it depends on a
modified version os meta-agl-lge (which include boost as dependency) to
build properly. This integration is still very primitive and will be
improved soon once we have WAM build system ported from qmake to CMake.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Remove QVariant leftovers

- This patch removes QVariant usage from PalmSystem* and
WebAppManagerConfig classes.
- Completely remove PalmSystemBase::getResource() and
WebAppManagerConfig::get/setConfiguration() since they are not being
used anywhere else.
- Modify PalmSystemWebOS's pmLog* functions that were expecting
QVariants as parameter to receive std::strings instead, since they were
being used only with QStrings.
- Anyway, if we really need those methods/variables in the future we can
use boost::any to replace QVariant basic functionality.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Replace QLocale by Boost.Locale

- This patch removes QLocale usage from WAM. It was being
used in WebPageBlink, to generate search path for localized
error pages.
- The previous implementation assumed QLocale::bcp47Name()
function returns an IETF Language Tag (BCP 47) string for
language string got from DeviceInfo API. Even though after
some tests with Qt API, it seems that bcp47Name() is not being
including 'script' field even when it's passed in the input string,
this path implements a boost-based genPathForLang() function that
returns a similar formatted path assuming the input string follows
the format decribed in QLocale documentation [1]

[1] http://doc.qt.io/qt-5/qlocale.html#QLocale-1

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Trim timestamp string in NetworkStatus

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Use const char arrays instead of macros for WindowTypes

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Remove Qt dependency from qmake files

For now, Qt::Core is kept as dependency on non-AGL (WebOSOSE) build os
WAM. That dependency will be removed in a followup patch.

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Get rid of Qt usage in webos/DeviceInfoImpl

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Port to CMake

TODO: Verify/fix webos build

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

jaragunde commented:
>     * To get this branch building correctly, a modified version of **meta-agl-lge** which adds _boost_ as WAM dependency. A PR has been sent to **meta-agl-lge** repository and its merge must be coordinated with this PR.

Chromium and WAM recipes are now part of meta-agl-lge, the patch to add boost should be sent there.

nickdiego commented:
> Chromium and WAM recipes are now part of meta-agl-lge, the patch to add boost should be sent there.

Hi jacobo, thanks for pointing out this. I guess you meant `meta-agl-devel`, right? I already have patches ready to send to review on AGL's gerrit at https://github.com/nickdiego/meta-agl-devel/tree/qtless-cmake

_PS: Updated the PR description._

nickdiego committed:
[agl][qtless] Fix error handling when converting strings to numbers

- Use boost-backed StringUtil's stringTo template function to
reliably convert strings to number types. Error handling follows
the same approach of Qt's QString::toInt(), using an optional
argument to hold the result status for the conversion operation.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
[agl][qtless] Use unordered_map instead of std::map

Since at this point no QObject is used as std::map key,
it's safe and easier fully migrate to more efficient
std::unordered_{map,multimap}.

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

jaragunde commented on src/util/LogMsgId.h:
```diff
@@ -17,7 +17,7 @@
 #define LOGMSGID_H
 
-#include <qglobal.h>
+#include <QtGlobal>
```
A leftover here?

nickdiego replied:
Good catch! I'll remove it. Thanks

jaragunde commented:
Tested on Minnowboard and R-Car M3 without issues.

jaragunde commented:
Also tested on a Raspberry Pi 3.

About the missing APs:
* WebOS Build (@jdapena?)
* WebOS regression tests (@jdapena / @jaragunde)

These require more work as we would have to remove Qt from the WebOS-specific code paths. I propose to proceed with the merge into the AGL branch, unless others have more comments, and continue the Qt-less work after that, as part of the upstreaming of AGL changes.

nickdiego commented:
> I propose to proceed with the merge into the AGL branch, unless others have more comments, and continue the Qt-less work after that, as part of the upstreaming of AGL changes.

Agreed.

jdapena commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,357 @@
+if(IS_WEBOS)
+    set(CMAKE_AUTOMOC ON)
+    if(USE_LUNA_SERVICE)
+        pkg_search_module(LIBLUNAPREFS REQUIRED luna-prefs)
```
(Can be fixed later, when adding support for building with cmake in webOS): this is missing luna-service2 dependency, that is something different from luna-prefs. It can be detected with pkg-config luna-service2

jdapena reviewed (approved):
It is good (for me) for landing in AGL branch. It will likely need more work (testing, etc), for landing in webOS OSE branches. But that can happen later.
